### DISCUSSION TITLE
Non deterministic behaviour of deterministic samplers with IPEX

### DISCUSSION THREAD

**Discussion by mstuermer** [vladmandic/sdnext#873] [Q&A] [answered]
Hello, 

I am running this fork under Gentoo Linux, kernel version 6.2.14, with Intel ARC 770 GPU and Intel ONE API successfully enabled. I am launching the script with "./webui.sh --use-ipex", Torch is detecting the GPU correctly and using it. Iteration speed is around 6 it/s. CPU is AMD Ryzen 7 3800X, 32 GB RAM, BAR is enabled, Python 3.10 is being used. No major errors/warnings on launch of the fork are given, so far I am using the default settings (aside theme). Image creation works, yeeaaaah! 

So far, so good. Now to my question: I am experiencing non deterministic behaviour with deterministic samplers, meaning even when giving the exact same prompt, seed and CFG scale using the same seed, this always results in totally different pictures. So in short pressing the generate button twice without changing anything results in two very different pictures with UniPC. 

Therefore my question: what might be the reason for that behaviour? Is it using IPEX, so the GPU acceleration, or might there be other switches enabled in the fork by default which could causes this? If the latter, which settings should I take a look at? Thanks in advance! 

The pics below show the result of running exact the same parameters two times, one after the other. 

![sd1](https://github.com/vladmandic/automatic/assets/1525998/d55dcf56-9c39-40a8-af6b-230cb9d22804)
![sd2](https://github.com/vladmandic/automatic/assets/1525998/a287cb25-6fa4-4c6b-891b-305add1ace50)

**Comment by vladmandic** [comment]
two things come to mind:
- some random number generators can result in non-deterministic sequence even if its initialized using same rng-seed. that may be the case with ipex implementation, i don't know
- most cross-attention optimizations are non-deterministic by default as it provides a touch more performance. which are you using?

**Comment by mstuermer** [comment]
Thanks for your quick reply!

I am using the Scaled-Dot-Product cross-attention optimazation at the moment, no advanced options on that enabled.

**Reply by vladmandic** [reply]
SDP by default is non-deterministic, there is a sub-option just underneath to enable deterministic mode for it.

**Comment by mstuermer** [comment]
Do you mean the button marked in red, "SDP disable memory attention?" This lead to no change in the behaviour. Also when using other Cross-attention optimization methods or disable it the observed behaviour is still the same. 

![sd3](https://github.com/vladmandic/automatic/assets/1525998/62684be5-30c7-4ca9-a719-7cf8164270a6)

**Comment by vladmandic** [comment, marked as answer]
yes, that's the button i meant - that makes SDP deterministic. but that still leaves the first problem of how is random generator behaving.
only way to go around that is not to use random generator provided by gpu libraries and use cpu random generator, but i really don't want to go down that path.

**Comment by Mindset-Official** [comment]
I added this to devices.py get the seed to reproduce, it won't work on batches you have to set the random seed to .01, 

def randn(seed, shape):

    if shared.cmd_opts.use_ipex:
    	torch.manual_seed(seed), torch.xpu.manual_seed_all(seed)
    	
    else:
    	torch.manual_seed(seed)
  #
    if device.type == 'mps':
        return torch.randn(shape, device=cpu).to(device)
    return torch.randn(shape, device=device)

I don't actually know how to code but I took a look at another ipex repo and tried different stuff, Not sure if this would interfere with other devices etc.

**Reply by vladmandic** [reply]
@Mindset-Official thanks, i'll something on those lines.
